File: TouchNothing/Utilities/LeaderboardRowMapper.swift
import Foundation

enum LeaderboardRowMapper {
    static func map(
        entries: [LeaderboardEntry],
        currentNickname: String?
    ) -> [LeaderboardRow] {
        mapRanked(
            items: entries.map { ($0.id.uuidString, $0.nickname, $0.durationSeconds) },
            currentNickname: currentNickname
        )
    }

    static func mapTotal(
        entries: [TotalLeaderboardEntry],
        currentNickname: String?
    ) -> [LeaderboardRow] {
        mapRanked(
            items: entries.map { ($0.nickname, $0.nickname, $0.totalDurationSeconds) },
            currentNickname: currentNickname
        )
    }

    private static func mapRanked(
        items: [(id: String, nickname: String, durationSeconds: Int)],
        currentNickname: String?
    ) -> [LeaderboardRow] {
        let normalizedCurrent = currentNickname.map { NicknameValidator.normalized($0).lowercased() }

        return items.enumerated().map { index, item in
            LeaderboardRow(
                id: item.id,
                rank: index + 1,
                nickname: item.nickname,
                durationSeconds: item.durationSeconds,
                isCurrentUser: item.nickname.lowercased() == normalizedCurrent
            )
        }
    }
}

File: app/src/main/java/com/touchnothing/app/domain/LeaderboardRowMapper.kt
package com.touchnothing.app.domain

import com.touchnothing.app.data.model.LeaderboardEntry
import com.touchnothing.app.data.model.LeaderboardRow
import com.touchnothing.app.data.model.TotalLeaderboardEntry
import com.touchnothing.app.util.NicknameValidator

object LeaderboardRowMapper {
    fun map(entries: List<LeaderboardEntry>, currentNickname: String?): List<LeaderboardRow> =
        mapRanked(
            items = entries.map { Triple(it.id, it.nickname, it.durationSeconds) },
            currentNickname = currentNickname,
        )

    fun mapTotal(entries: List<TotalLeaderboardEntry>, currentNickname: String?): List<LeaderboardRow> =
        mapRanked(
            items = entries.map { Triple(it.nickname, it.nickname, it.totalDurationSeconds) },
            currentNickname = currentNickname,
        )

    private fun mapRanked(
        items: List<Triple<String, String, Int>>,
        currentNickname: String?,
    ): List<LeaderboardRow> {
        val normalizedCurrent = currentNickname?.let {
            NicknameValidator.normalized(it).lowercase()
        }
        return items.mapIndexed { index, item ->
            LeaderboardRow(
                id = item.first,
                rank = index + 1,
                nickname = item.second,
                durationSeconds = item.third,
                isCurrentUser = item.second.lowercase() == normalizedCurrent,
            )
        }
    }
}
